src/deepsparse/server/main.py

```diff
@@ -0,0 +1,212 @@
+# limitations under the License.
+
+"""
+
+"""
+
+import logging
+from pathlib import Path
+
+import click
+
+from deepsparse.server.asynchronous import execute_async, initialize_aysnc
+from deepsparse.server.config import (
+    ServerConfig,
+    server_config_from_env,
+    server_config_to_env,
+)
+from deepsparse.server.pipelines import load_pipelines_definitions
+from deepsparse.server.utils import serializable_response
+from deepsparse.version import version
+
+
+try:
+    import uvicorn
+    from fastapi import FastAPI
+    from starlette.responses import RedirectResponse
+except Exception as err:
+    raise ImportError(
+        "Exception while importing install dependencies, "
+        "run `pip install deepsparse[server]` to install the dependencies. "
+        f"Recorded exception: {err}"
+    )
+
+
+ENV_DEEPSPARSE_SERVER_CONFIG = "DEEPSPARSE_SERVER_CONFIG"
+_LOGGER = logging.getLogger(__name__)
+
+
+def _add_general_routes(app, config):
+    @app.get("/config", tags=["general"], response_model=ServerConfig)
+    def _info():
+        return config
+
+    @app.get("/ping", tags=["general"])
+    @app.get("/health", tags=["general"])
+    @app.get("/healthcheck", tags=["general"])
+    def _health():
+        return {"status": "healthy"}
+
+    @app.get("/", include_in_schema=False)
+    def _home():
+        return RedirectResponse("/docs")
+
+    _LOGGER.info("created general routes, visit `/docs` to view available")
+
+
+def _add_pipeline_route(app, pipeline_def, num_models: int, defined_tasks: set):
+    async def _predict_func(request: pipeline_def.request_model):
+        results = await execute_async(
+            pipeline_def.pipeline, **vars(request), **pipeline_def.kwargs
+        )
+        return serializable_response(results)
+
+    if pipeline_def.config.alias:
+        # add the prediction path under the given alias
+        app.post(
+            f"/predict/{pipeline_def.config.alias}",
+            response_model=pipeline_def.response_model,
+            tags=["prediction"],
+        )(_predict_func)
+    elif num_models == 1:
+        # add only a single predict route since this is the only model we're serving
+        app.post(
+            "/predict", response_model=pipeline_def.response_model, tags=["prediction"]
+        )(_predict_func)
+    elif pipeline_def.config.task not in defined_tasks:
+        # fall back on adding the model to the task provided nothing is already assigned
+        app.post(
+            f"/predict/{pipeline_def.config.task}",
+            response_model=pipeline_def.response_model,
+            tags=["prediction"],
+        )(_predict_func)
+        defined_tasks.add(pipeline_def.config.task)
+    else:
+        raise ValueError(
+            f"Multiple tasks defined for {pipeline_def.config.task} and no alias "
+            f"given for {pipeline_def.config}. "
+            "Either define an alias or supply a single model for the task"
+        )
```

Comment:
since we are now only adding one route per `pipeline_def` with the only difference between them being the route name, it would probably be cleaner to choose the route first (raise error if applicable), then define+decorate the predict fn

Reply:
yep, good catch